craft code refactor

Vallat opened the pull request:
## What Does This PR Do
This is port of [my PR](https://github.com/ss220-space/Paradise/pull/548) made a long time ago
This PR changes how personal crafting calculates used materials for deletion, fixing possible infinite-loop (we caught it several times and it crashed the server) and incorrect logic for using multiple stackable items (cables, reagent containers, etc.)

## Why It's Good For The Game
Less bugs more code readability. No issues noticed since 5 months of being on live servers

## Changelog
:cl:
fix: Personal crafting now uses materials correctly
/:cl:

lewcc commented:
Early returns my beloved

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,125 @@
+	if(!parts)
+		return ", missing component."
+					
+	var/atom/movable/craft_result = new recipe.result (get_turf(user.loc))
```
recipe.result is a list of typepaths. Does `new` really meaningfully work on that? And even if does work, i'd expect it to create a new *list* of objects, not a single `/atom/movable`?

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,125 @@
+	return 0
+			
+
+/datum/personal_crafting/proc/del_reqs(datum/crafting_recipe/recipe, mob/user)
```
the old comment is terrible of course, but it'd be nice to replace it with *some* english description of what the proc does. The name is rather cryptic. It deletes.. req..requirements?.. What does it mean, and why would you want to do that?

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,125 @@
+				if(!contained_reagent)
+					continue
+				
+				if(contained_reagent.volume > needed_amount)
+					things_for_deletion[container] = list(contained_reagent, needed_amount)
+					part_reagent.volume = needed_amount
+					part_reagent.data = contained_reagent.data
+					needed_amount = 0
+					break
+				else
+					needed_amount -= contained_reagent.volume
+					things_for_deletion[container] = list(contained_reagent, contained_reagent.volume)
+					part_reagent.volume += contained_reagent.volume
+					part_reagent.data += contained_reagent.data
```
two if branches are nearly identical. I'd go for something more unified like

```dm
var/extracted_amount = min(contained_reagent.volume, needed_amout)
things_for_deletion[container] = list(contained_reagent, extracted_amount)
part_reagent.volume += extracted_amount
part_reagent.date += contained_reagent.data
needed_amount -= extracted_amount
if(needed_amount <= 0)
  break
```

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,125 @@
+				if(!istype(item_stack, thing))
+					continue
+				
+				if(item_stack.amount > needed_amount)
```
same deal with this `if`

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,125 @@
+		else if(ispath(thing, /obj/item/stack))
+			var/obj/item/stack/part_stack = locate(thing) in parts_list_for_return
+			if(!part_stack)
+				part_stack = new thing()
+				part_stack.amount = 0
```
```suggestion
				part_stack = new thing(loc = null, new_amount = 0)
```
(at least i'd think that'd work)

Vallat replied:
Actually didn't work because almost all items just don't have `new_amount` argument in their overriden `New()`

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,125 @@
+					part_reagent.data += contained_reagent.data
+	
+			if(needed_amount > 0)
+				return null
```
If we reach this point, something has gone wrong (because we must have CHECKED beforehad that we have enough stuffs!), and problems probably arose (we already deleted reagents/stacks, while iterating through previous reqs).
This is a kind of "this should never happen" situation in which i'd think a runtime would be appropriate.

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,125 @@
+			parts_list_for_return += part_atom
+
+	for(var/thing in things_for_deletion)
+		if(istype(thing, /obj/item/reagent_containers))
```
You put the reagents and the stacks into the same list, but then istype the items when iterating, to differentiate them back, and take completely separate code paths for both.
I think having two lists would be easier to follow.

Or even: handle the deletion right there in the original loop where you populate the list(s).

moxian reviewed (changes requested):
I checked out the pr, tried crafting black carpet.
Spawned the carpet, spawned a black crayon, used the construction menu to construct.
Carpet was consumed, black carpet was not produced. Did you test the code?

Vallat commented:
Thanks for the review, I really just didn't notice that recipe `result` is list here and that's why item spawning was not working
Fixed it and changed some code according to your suggestions

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,121 @@
+	
+	for(var/possible_result in recipe.result)
+		var/atom/movable/craft_result = new possible_result (get_turf(user.loc))
+		craft_result.CheckParts(parts, recipe)
```
This is a pre-existing problem, not your code, but.. why are we even callign CheckParts here?
Note, that CheckParts does not strictly speaking "check" any parts, instead it *moves* the parts into the newly created thing. 
This leads to problems where repeatedly crafting and un-crafting, say, security HUDsunglasses makes HUDsunglasses that contain security HUD that contains security HUDsunglasses that contain securityt HUD that contain...

Does this serve any purpose whatsoever? We can't extract the contents from the resulting item anyway.. Or can we?.. Or is this essential to making foods provide nutrition (i don't know)?..

Vallat replied:
¯\_(ツ)_/¯
I'm assuming this was done for future features, like inspecting what items are made of or something like that.

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,121 @@
+					break
+	
+			if(needed_amount > 0)
+				stack_trace("Bad things happened in [__FILE__] at [__LINE__], we didn't have as many reagents as we expected!")
```
```suggestion
				stack_trace("While crafting [recipe], some of [thing] went missing (still need [needed_amount])!")
				continue // ignore the error, and continue crafting for player's benefit
```
stacktrace already gives the filename and the line, but it *doesn't* give the locals.

Vallat replied:
If I remember correctly `stack_trace` gives the filename and the line where `stack_trace` located, but not the actual place where the runtime error happened

moxian replied:
```
[2022-05-09T07:06:45] Starting up. Round ID is NULL
[2022-05-09T07:06:45] -------------------------
[2022-05-09T07:06:45] Running ParaCode commit: bdfac98376220733305d7d7af6da04c733c04462 (Date: 2022-05-08)
[2022-05-09T07:08:49] Runtime in unsorted.dm,1932: hello
[2022-05-09T07:08:49]   proc name: stack trace (/datum/proc/stack_trace)
[2022-05-09T07:08:49]   usr: Ellie Beach (moxian) (/mob/living/carbon/human)
[2022-05-09T07:08:49]   usr.loc: The shuttle floor (40,143,2) (/turf/simulated/floor/mineral/titanium/blue)
[2022-05-09T07:08:49]   src: /datum/personal_crafting (/datum/personal_crafting)
[2022-05-09T07:08:49]   call stack:
[2022-05-09T07:08:49]   /datum/personal_crafting (/datum/personal_crafting): stack trace("hello")
[2022-05-09T07:08:49]   /datum/personal_crafting (/datum/personal_crafting): requirements deletion(Security HUD removal (/datum/crafting_recipe/hudsunsecremoval), Ellie Beach (/mob/living/carbon/human))
[2022-05-09T07:08:49]   /datum/personal_crafting (/datum/personal_crafting): construct item(Ellie Beach (/mob/living/carbon/human), Security HUD removal (/datum/crafting_recipe/hudsunsecremoval))
[2022-05-09T07:08:49]   /datum/personal_crafting (/datum/personal_crafting): ui act("make", /list (/list), /datum/tgui (/datum/tgui), /datum/ui_state/not_incapacita... (/datum/ui_state/not_incapacitated_state))
[2022-05-09T07:08:49]   /datum/tgui (/datum/tgui): Topic("src=%5B0x21017550%5D_679&actio...", /list (/list))
[2022-05-09T07:08:49]   Moxian (/client): Topic("src=%5B0x21017550%5D_679&actio...", /list (/list), /datum/tgui (/datum/tgui))
```
While I guess it's true that there's no filename/line number here, there is proc name and it should really be enough.

Vallat replied:
Alright then

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,121 @@
+					break
+			
+			if(needed_amount > 0)
+				stack_trace("Bad things happened in [__FILE__] at [__LINE__], we didn't have as many items as we expected!")
```
```suggestion
				stack_trace("While crafting [recipe], some of [thing] went missing (still need [needed_amount])!")
				continue
```

moxian commented on code/modules/crafting/craft.dm:
```diff
@@ -151,152 +151,121 @@
-		Deletion.Cut(Deletion.len)
-		qdel(DL)
+			var/atom/movable/part_atom = locate(thing) in (surroundings - parts_list_for_return)
+			if(!part_atom)
+				stack_trace("Bad things happened in [__FILE__] at [__LINE__], we didn't manage to find the atom we need for crafting!")
```
```suggestion
			if(!part_atom)
				stack_trace("While crafting [recipe], the [thing] went missing!")
				continue
```

moxian reviewed (approved):
LGTM modulo runtimes wording.

TrashDoxx commented:
When?
